### PR TITLE
DNS Zones: List / Get

### PR DESCRIPTION
**Pull request by jtopjian**
For #172 

Supersedes #173 

Here's some more info on the DNS methods: https://github.com/openstack/designate/blob/stable/newton/designate/api/v2/controllers/zones/__init__.py

And on the schema: 

https://github.com/openstack/designate/blob/stable/newton/designate/objects/zone.py
https://github.com/openstack/designate/blob/6ddf20aba5c361629168ce2f3103c1e8052528d5/designate/objects/adapters/base.py

### CODE REVIEW

**Comment by jtopjian** [comment]
I'm still working on this. Unit tests pass, but I want to get some basic acceptance tests first just to make sure things work correctly against a running Designate service.

The first commit is just a rebase to get things up to date with master. The other commits fix things up to get the unit tests to pass.

I'm going to add some inline comments with a few questions.

File: openstack/client.go [hunk 1]
```diff
@@ -317,7 +317,10 @@ func NewDNSV2(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts) (
 	if err != nil {
 		return nil, err
 	}
-	return &gophercloud.ServiceClient{ProviderClient: client, Endpoint: url}, nil
+	return &gophercloud.ServiceClient{
+		ProviderClient: client,
+		Endpoint:       url,
+		ResourceBase:   url + "v2/"}, nil
```
Comment (jtopjian) [inline openstack/client.go]: @jrperritt is this the correct thing to do? I noticed this was done for some services, but not all. I haven't dug into why (been a little pressed for time) but will do later just so I see why.

File: openstack/dns/v2/zones/results.go [hunk 2]
```diff
@@ -90,6 +93,7 @@ func (r *Zone) UnmarshalJSON(b []byte) error {
 		CreatedAt     gophercloud.JSONRFC3339MilliNoZ `json:"created_at"`
 		UpdatedAt     gophercloud.JSONRFC3339MilliNoZ `json:"updated_at"`
 		TransferredAt gophercloud.JSONRFC3339MilliNoZ `json:"transferred_at"`
+		Serial        interface{}                     `json:"serial"`
```
Comment (jtopjian) [inline openstack/dns/v2/zones/results.go]: Without this, I was getting scientific notations as described in #209. I saw this was done for `compute/v2/flavors` for `Swap`, so implemented it here.

Is this the best way to go? Or should this wait until a solution in #209 is implemented?

**Comment by jtopjian** [comment]
@jrperritt There looks to be an issue with `AllPages`. In the unit tests, it's returning a count of 3 consistently. In the acceptance test, it's randomly returning 0 or 3. If 3, one of the zones is entirely blank.

Other than that, and with the other comments in mind, this is ready for a first review. :)

**Comment by jrperritt** [comment]
@jtopjian Please rebase with master to pull in #276 and see if that fixes the `AllPages` issue.

**Comment by jtopjian** [comment]
Working great :)

**Comment by jtopjian** [comment]
@jrperritt ping for review

**Comment by jtopjian** [comment]
Just noticed ListOpts was not implemented. _Now_ it should be ready for a first review.

File: acceptance/openstack/dns/v2/zones.go [hunk 3]
```diff
@@ -0,0 +1 @@
+package v2
```
Comment (jrperritt) [inline acceptance/openstack/dns/v2/zones.go]: This file is by convention just named `pkg.go` or `doc.go`.

File: openstack/client.go [hunk 4]
```diff
@@ -310,6 +310,19 @@ func NewDBV1(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts) (*
...
+func NewDNSV2(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts) (*gophercloud.ServiceClient, error) {
+	eo.ApplyDefaults("dns")
+	url, err := client.EndpointLocator(eo)
+	if err != nil {
+		return nil, err
+	}
+	return &gophercloud.ServiceClient{
+		ProviderClient: client,
+		Endpoint:       url,
+		ResourceBase:   url + "v2/"}, nil
```
Comment (jrperritt) [inline openstack/client.go]: This is for when the catalog's URL for this service doesn't include the version; it occurred with Networking v2 I think. Is that also the case here?

File: acceptance/openstack/dns/v2/zones.go (hunk 3 above)
Comment (jtopjian) [inline acceptance/openstack/dns/v2/zones.go]: Ah - OK, thanks.

If it's OK with you, this file gets renamed to `dns.go` in the next PR which follows the pattern of the other acceptance tests I've done. It also has actual contents.

No issues renaming it to `pkg.go` in this PR if you prefer, though.

File: openstack/client.go (hunk 4 above)
Comment (jtopjian) [inline openstack/client.go]: Got it. And that's correct: the catalog I'm testing against does not have a version in the URL.

File: acceptance/openstack/dns/v2/zones.go (hunk 3 above)
Comment (jrperritt) [inline acceptance/openstack/dns/v2/zones.go]: Ah, no problem. The name of the file isn't really important.

File: openstack/client.go (hunk 4 above)
Comment (jrperritt) [inline openstack/client.go]: :+1:

**Comment by jrperritt** [comment]
Looks good

**Pull request merged**